sled-agent/src/common/instance.rs

```diff
@@ -303,78 +288,11 @@ impl InstanceStates {
-                        self.switch_propolis_id_to_target(observed.time);
-
-                        assert_eq!(self.propolis_role(), PropolisRole::Retired);
-                    }
-
-                    // This is a successful migration in. Point the instance to the
-                    // target VMM and clear migration IDs so that another migration
-                    // in can begin. Propolis will continue reporting that this
-                    // migration was successful, but because its ID has been
-                    // discarded the observed migration status will change from
-                    // Succeeded to NoMigration.
-                    //
-                    // Note that these calls increment the instance's generation
-                    // number twice. This is by design and allows the target's
-                    // migration-ID-clearing update to overtake the source's update.
-                    PropolisRole::MigrationTarget => {
-                        self.switch_propolis_id_to_target(observed.time);
-                        self.clear_migration_ids(observed.time);
-
-                        assert_eq!(self.propolis_role(), PropolisRole::Active);
-                    }
-
-                    // This is a migration source that previously reported success
-                    // and removed itself from the active Propolis position. Don't
-                    // touch the instance.
-                    PropolisRole::Retired => {}
-                }
-            }
-            ObservedMigrationStatus::Failed => {
-                self.transition_migration(
-                    MigrationState::Failed,
-                    observed.time,
-                );
-
-                match self.propolis_role() {
-                    // This is a failed migration out. CLear migration IDs so that
-                    // Nexus can try again.
-                    PropolisRole::Active => {
-                        self.clear_migration_ids(observed.time);
-                    }
-
-                    // This is a failed migration in. Leave the migration IDs alone
-                    // so that the migration won't appear to have concluded until
-                    // the source is ready to start a new one.
-                    PropolisRole::MigrationTarget => {}
-
-                    // This VMM was part of a failed migration and was subsequently
-                    // removed from the instance record entirely. There's nothing to
-                    // update.
-                    PropolisRole::Retired => {}
-                }
-            }
-            ObservedMigrationStatus::InProgress => {
-                self.transition_migration(
-                    MigrationState::InProgress,
-                    observed.time,
-                );
-            }
-            ObservedMigrationStatus::NoMigration
-            | ObservedMigrationStatus::Pending => {}
+        if let Some(m) = observed.migration_in {
```

Comment:
Would it ever be possible to have `migraiton_in` and `migration_out` set at the same time?

Reply:
Yes, see oxidecomputer/propolis#661: Propolis will continue to report the
migration in forever once an instance has been migrated in. In practice, it seems
remarkably unlikely for a sled-agent to see a *changed* migration in state for
the first time in an update that also includes a migration out, but we may as
well handle it correctly.